velox/exec/tests/MergeJoinTest.cpp

```diff
@@ -45,6 +45,87 @@ class MergeJoinTest : public HiveConnectorTestBase {
     return params;
   }
 
+  std::vector<RowVectorPtr> generateInput(const std::vector<VectorPtr>& keys) {
+    std::vector<RowVectorPtr> data;
+    data.reserve(keys.size());
+    vector_size_t startRow = 0;
+
+    for (const auto& key : keys) {
+      auto payload = makeFlatVector<int32_t>(
+          key->size(), [startRow](auto row) { return (startRow + row) * 10; });
+      data.push_back(makeRowVector({key, payload}));
+      startRow += key->size();
+    }
+    return data;
+  }
+
+  // Lazy vector loader class to ensure they get loaded in the correct order,
+  // and only once.
+  class MySimpleVectorLoader : public VectorLoader {
+   public:
+    explicit MySimpleVectorLoader(
+        size_t batchId,
+        const std::shared_ptr<size_t>& count,
+        std::function<VectorPtr(RowSet)> loader)
+        : batchId_(batchId), count_(count), loader_(loader) {}
+
+    void loadInternal(
+        RowSet rows,
+        ValueHook* hook,
+        vector_size_t resultSize,
+        VectorPtr* result) override {
+      if (batchId_ > *count_) {
+        *count_ = batchId_;
+      }
+      VELOX_CHECK_GE(batchId_, *count_, "Lazy vectors loaded out of order.");
```

Comment:
This check should be before we update `count_`, no?

Reply:
It doesn't really matter, the effect should be the same.